Amazon.Ion.ObjectMapper/IonDictionarySerializer.cs

```diff
@@ -0,0 +1,52 @@
+﻿using Amazon.IonDotnet;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace Amazon.Ion.ObjectMapper
+{
+    public class IonDictionarySerializer : IonSerializer<IDictionary>
+    {
+        private IonSerializer serializer;
+        private Type valueType;
+
+        public IonDictionarySerializer(IonSerializer ionSerializer, Type valueType)
+        {
+            this.serializer = ionSerializer;
+            this.valueType = valueType;
+        }
+
+        public override IDictionary Deserialize(IIonReader reader)
+        {
+            reader.StepIn();
+
+            Type typedDictionaryType = typeof(Dictionary<,>).MakeGenericType(typeof(string), this.valueType);
+            MethodInfo methodInfo = typedDictionaryType.GetMethod("Add");
+            var dictionary = Activator.CreateInstance(typedDictionaryType);
+            IonType currentType;
+            while ((currentType = reader.MoveNext()) != IonType.None)
+            {
+                object[] parameters = new object[] { 
+                    reader.CurrentFieldName, 
+                    Convert.ChangeType(this.serializer.Deserialize(reader, valueType, currentType), valueType) };
+                methodInfo.Invoke(dictionary, parameters);
+
+            }
+
+            reader.StepOut();
+            return (IDictionary)dictionary;
+        }
+
+        public override void Serialize(IIonWriter writer, IDictionary item)
```

Comment:
Should `IDictionary` be generic?

Reply:
Or - do we need methods for both non-generic and generic dictionaries?

Reply:
The object mapper should not support non-generic dictionaries. It's not really appropriate for something that maps to a struct object.

With that said, the type of the DictionarySerializer itself can't take a generic dictionary due to then needing a specific DictionarySerializer for all the supported value types. (including itself recursively) which be a huge amount of repetition for no value.

Using generic IDictionary gives us some functionality and casting compatibility that we need.

The List Serializer is designed in the same way, which now I understand why.

Reply:
I'm not sure I understand the comment. By non-generic did you mean generic (aka typed)? Non-generic would be `IDictionary`, which this method accepts already...

Reply:
To answer the question succinctly, the current implementation is non-generic and it covers the use case for generic dictionaries so there's no need for the generic dictionaries. If we did use generic dictionaries, it would require us to have basically a list of Dictionaries for each primitive type we support.